src/librustc_typeck/check/method/suggest.rs

```diff
@@ -61,12 +64,62 @@ pub fn report_error<'a, 'tcx>(fcx: &FnCtxt<'a, 'tcx>,
                 None);
 
             // If the item has the name of a field, give a help note
-            if let (&ty::TyStruct(did, _), Some(_)) = (&rcvr_ty.sty, rcvr_expr) {
+            if let (&ty::TyStruct(did, substs), Some(expr)) = (&rcvr_ty.sty, rcvr_expr) {
                 let fields = ty::lookup_struct_fields(cx, did);
-                if fields.iter().any(|f| f.name == item_name) {
-                    cx.sess.span_note(span,
-                        &format!("use `(s.{0})(...)` if you meant to call the \
-                                 function stored in the `{0}` field", item_ustring));
+
+                if let Some(field) = fields.iter().find(|f| f.name == item_name) {
+                    let expr_string = match cx.sess.codemap().span_to_snippet(expr.span) {
+                        Ok(expr_string) => expr_string,
+                        _ => "s".into() // Default to a generic placeholder for the
+                                        // expression when we can't generate a string
+                                        // snippet
+                    };
+
+                    macro_rules! span_stored_function {
+                        () => {
+                            cx.sess.span_note(span,
+                                              &format!("use `({0}.{1})(...)` if you meant to call \
+                                                        the function stored in the `{1}` field",
+                                                       expr_string, item_ustring));
+                        }
+                    };
+
+                    macro_rules! span_did_you_mean {
+                        () => {
+                            cx.sess.span_note(span, &format!("did you mean to write `{0}.{1}`?",
+                                                             expr_string, item_ustring));
+                        }
+                    };
```

Comment:
I forgot to ask before pushing these changes, is it alright to use a macro here? Is a closure preferred?